### ISSUE TITLE
Add PUT support for saving a file

### ISSUE THREAD

**Issue by cfry** [kzahel/web-server-chrome#61]
I'm building an IDE for JavaScript. I want the user to be able to save their code
in an ordinary ascii file in their choice of location in a file under the
root of the server. Saving to the same location should replace the
contents of the file previously at that location.
The saved location should have an optional folder (or nested folder),
a file name and a file extension like .js, .html, or .txt
The capability to save non-ascii files like jpegs, audio, etc would be nice.

**Comment by kzahel** [comment]
Cool. I'll look into adding this soon. I look forward to checking out your
app

**Comment by cfry** [comment]
Thanks Kyle,
I'm impressed with your web server. Its got to be the easiest to set up web server ever.

I'm building an IDE for programming robots to make things
   (think "more advanced 3D printers" and you'll have a good idea.)
The startup I'm with hasn't decided to release it yet but certainly
your feedback would be great so I'll let you know when its out.

I'm using Google's "Chrome Apps" platform which has a very restricted
way to access files. Your tool gives me that extra functionality that
programmers really need.

I'd also like to be able to get a directory listing from your server.
I note that I can do this manually from a browser by just entering in
the folder. But for some reason I don't yet understand,
when I try to do this programmatically from within the Chrome app,
I get back a page that looks like its from Google and it doesn't have
the HTML that shows what files are in the folder. Maybe Google is
blocking this somehow? I can imagine it being a "security restriction"
though I can't figure out how they would block it.
But what I really want is not a page of HTML that I'd have to screen scrape
to get back the file list, I just want the raw list of contained file names,
newline separated.
If you were to do this, I imagine the url requested would be something like
"/some-folder/.../directory_listing". Perhaps there should be an option in your set-up screen
as this is somewhat of a security risk, esp. combined with the
"PUT" of my original suggestion, ie malicious code could write the empty string
into your files for instance. I hate this security aspect to modern programming
as it makes everything so much harder, but at least with the proper options,
a user decides for themselves what risk they can tolerate.

Anyway, great tool and well worth your impressive 28K users!
________________________________

**Comment by kzahel** [comment]
I'll provide a more API like access to just return a JSON response of directory listings. I'll keep you updated.
Right now you can append ?static=1 to get static html. It's actually already serving the directory listing as a js array that the page renders using javascript...

**Comment by kzahel** [comment]
Maybe I'll try to copy an API similar to say google drive or dropbox HTTP api

**Comment by cfry** [comment]
Aha, that ?static=1  suffix did the trick and I can get the HTML for a directory listing!
Looks nice and clean so not hard to parse,
but yes, JSON would be nice.

Thanks tons!

PS: Here's one thing that could save you time in support:
have a doc page that has a link to it in the Web Server for Chrome
interface itself.
Say at the top of that page: "No, it doesn't do PHP!"


________________________________

**Comment by kzahel** [comment]
Just added support
https://github.com/kzahel/web-server-chrome/commit/9acb5ae93ca6ec13057bb61b0686222904f88e83

Will maybe do a bit more testing and then put on web store. It's an option under advanced options.

**Comment by kzahel** [comment]
the way it works is you do a PUT request, where the URI is the same as it would be for a GET request, e.g. if the file is /testFolder/testfile.txt, you would make a PUT to the same url, and the request body is the binary data of the file. No need to set any other headers.

If the option is turned on you can drag and drop into the directory listing served by the web server and it will upload to the currently displayed folder, and if it works correctly, it will refresh teh page.

**Comment by kzahel** [comment]
I only tested with small files. I just tried now with a movie (700MB) and it crashes the tab. I think because it runs out of memory. I will need to come up with a streaming upload option...

**Comment by cfry** [comment]
Wow, that was fast! Thanks Kyle. It will significantly increase the flexibility
of my software AND yours.
A few comments below:
________________________________

**Comment by cfry** [comment]
I don't expect to be uploading files nearly so large, at least in the near term.
But I suppose I might need the large size some day.
________________________________

**Comment by kzahel** [comment]
For a text file, that's correct, just the raw data from the text file will work perfectly.

**Comment by kzahel** [comment]
It also gives a 400 when the file already exists. I can change it to replace an existing file, too, or make that also configurable.

**Comment by kzahel** [comment]
I just realized you probably want it to replace the file if it already exists, so I changed the default to simply replace the file if it already exists
https://github.com/kzahel/web-server-chrome/commit/92f10d32a1147d74d9f38c5ffea19b83c620d3cc

**Comment by cfry** [comment]
Mostly I'd want:
If file doesn't exist, create and write it,
else overwrite the existing file
(and by that I mean, even if the old file is longer than the new,
the complete file when done is the length of the new content

But that indicates a few other things someone might want, ie
"does_file_x_exist?"
I suppose someone might want
file size, and last write date too.
Maybe there's a file_info command that could tell you
both whether the file existed or not, and if so,
its size and mod date.
You can get file size from just getting the file and seeing how long it is,
but I suppose if you're thinking of handling LARGE files,
someone might not want to get a file that's more than XMbytes, for instance.

If we had file_info, then no critical need to put options on PUT
for things like "don't write if it already exists because
you could get file_info before you were about to write.
Some file systems would support other meta data too like
a difference between create and mod date, author, etc.
I'd say just do what's easy and "universal" here so you don't have
to document different OS's.

________________________________

**Comment by cfry** [comment]
Yes that should certainly be the default.
________________________________

**Comment by kzahel** [comment]
There is already HEAD request support which is the standard way to check
whether file exists and it returns the size of the file as a content length
header

**Comment by cfry** [comment]
Great!
________________________________

**Comment by kzahel** [comment]
It also occurred to me that you might need a way to create directories too, right?

**Comment by cfry** [comment]
That hadn't occurred to me but yes, I guess that's true.
You could do: If a PUT request had its file name ending in slash,
then that "ensures" that the directory is there.
If such a request had content in the PUT, then consider it an error,
as that would imply that the caller thought they were making a file
not a dir.

There's the notion of:
"create the dirs you need"
which would mean create all the dirs in such a path if they weren't already
there. But I'd say that's not so safe, so
if the user attempted to
PUT  /foo/bar/baz/
and "bar" didn't exist, then error,
But if foo and bar existed, and baz didn't, create the dir "baz".
If baz exists, then its a no-op.

If you PUT /foo/bar/baz/frotz.txt
and baz didn't exist, I'd say error rather than create it.
As long as we can use HEAD to tell whether /foo/bar/baz/
exists, then a programmer can explicitly create
/foo/bar/baz/
and THEN create /foo/bar/baz/frotz.txt

________________________________

**Comment by kzahel** [comment]
Now support GET request to return JSON https://github.com/kzahel/web-server-chrome/commit/bf8d6da51c19f87fbe44082cfdaefb0a7203ab63
Simply specify ?json=1 parameter or send "Accept" header with application/json and it should return JSON instead of HTML

**Comment by derhuebiii** [comment]
Great and easy webserver in general, thanks! :)

How can I overwrite a file completely? Not only the beginning of the file if the existing file already exists and is longer?

**Comment by kzahel** [comment]
Ah, whoops. You're right. It looks like it maybe only replaces the first x bytes if the old file was x+y bytes. (not 100% sure, but it looks like that could be true). Is that what you're seeing?
I'll change the default behavior to remove old file first or truncate it. Maybe an extra option to allow writing to specific offset too

**Comment by kzahel** [comment]
https://github.com/kzahel/web-server-chrome/commit/e09a550ba44ac4938b4e633853b1eac71744d5f1 delete file now before putting new one in. haven't released to store yet though.

**Comment by ethanaobrien** [comment]
@kzahel Should this be closed?
